helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/HardConstraint.java

```diff
@@ -33,15 +34,46 @@ abstract class HardConstraint {
    * Check if the replica could be assigned to the node
    * @return True if the proposed assignment is valid; False otherwise
    */
-  abstract boolean isAssignmentValid(AssignableNode node, AssignableReplica replica,
+  abstract ValidationResult isAssignmentValid(AssignableNode node, AssignableReplica replica,
       ClusterContext clusterContext);
 
+
   /**
-   * Return class name by default as description if it's explanatory enough, child class could override
-   * the method and add more detailed descriptions
-   * @return The detailed description of hard constraint
+   * Stores the assignment validation result and the error message, in case of failure.
    */
-  String getDescription() {
-    return getClass().getName();
+  public static class ValidationResult {
```

Comment:
Why we need to have this kind of result? I am not fully convinced to have this structure. If a failure of a message require this new object. It will generated a lot of object during pipelines can cause GC.

Reply:
Yes, discussed offline. Changing the logic to just log the failed constraint (debug mode) and keeping the rest of the code as it is. Thanks Junkai!